### ISSUE TITLE
[sram_ctrl] Error out upon SRAM accesses during a key request.

### ISSUE THREAD

**Issue by msfschaffner** [lowRISC/opentitan#7059]
Currently, SRAM stalls all bus accesses while a key request is in progress. 
This can be problematic, since if EDN is not initialized, such a key request may stall the processor forever, even preventing it from entering an NMI (see the corresponding OTBN discussion https://github.com/lowRISC/opentitan/issues/7021).
A solution currently discussed for OTBN scrambling would be to make the bus access error out instead of stalling it.

**Comment by tjaychen** [comment]
these two situations felt a bit different to me. For the otbn, that's always programmed by ibex to go.
For the main sram scramble, I was originally imagining a scenario where live software is running and we need to "rotate" keys.

My original thinking is that we should stall, because of the situation is kind of like a cache invalidate.  We are waiting for the operation to finish before we can safely use it again.  This waiting could involve code that uses the stack / checking variables etc, hence why i thought maybe stalling would be better. 

@cfrantz 

You are right however stalling in general can be dangerous, but this kind of stall would not prevent for example the watchdog from eventually triggering.  I'm not against explicit erroring however, since most likely the software code around this area will have to be written in assembly anyways.

**Comment by tjaychen** [comment]
i should mention also because of this issue, we explicitly need to setup entropy inside rom, and ensuring the entropy complex can support this is something that's important.

**Comment by cfrantz** [comment]
I'm ok with SRAM returning aborts before the scrambling configuration is ready.

I assume that if I request a new scrambling key, I can let that happen "in the background" and execution will be fine as long as I don't touch SRAM before its ready?

Is there a status bit in the SRAM controller that I could spin on if I need to wait for SRAM:

Pseudocode:
```
abs_mmio_write32(SRAM_CTRL_REG, SCRAMBLING_GET_NEW_KEY_RIGHT_NOW);
while(abs_mmio_read32(SRAM_STATUS_REG) & REKEY_IN_PROGRESS)
  /* do nothing (or optionally: watchdog_pet())*/
  ;
```

**Comment by msfschaffner** [comment]
You should be able to spin on `sram_ctrl.STATUS.SCR_KEY_SEED_VALID` to achieve this. 
See also description of `sram_ctrl.CTRL.RENEW_SCR_KEY` in https://docs.opentitan.org/hw/ip/sram_ctrl/doc/index.html.

**Comment by mundaym** [comment]
Can we always safely ignore `sram_ctrl.STATUS.SCR_KEY_VALID` if we are checking `sram_ctrl.STATUS.SCR_KEY_SEED_VALID`?

**Comment by msfschaffner** [comment]
Sorry, I believe I have a copy paste error in what I have said in my previous comment.
You can poll on `sram_ctrl.STATUS.SCR_KEY_VALID` and safely ignore `sram_ctrl.STATUS.SCR_KEY_SEED_VALID` while polling.

**Comment by mundaym** [comment]
Thanks. Once `sram_ctrl.STATUS.SCR_KEY_VALID` is set what behavior do we want when `sram_ctrl.STATUS.SCR_KEY_SEED_VALID` is 0? Should we ignore it or should we trigger a shutdown?

**Comment by msfschaffner** [comment]
So it depends a bit how much knowledge of the provisioning and life cycle state ROM firmware has (or should have).

ROM needs to be able to run in both cases - especially during the test life cycle states where the key seeds have not been provisioned yet.

We could technically double check the seed valid status with the provisioning status of the corresponding OTP partition, and trigger a shutdown if a discrepancy is detected. This can be achieved by checking the digest value of the `SECRET1` partition in the [OTP CSRs](https://docs.opentitan.org/hw/ip/otp_ctrl/doc/#Reg_secret1_digest_0). If that one has a non-zero value, the `sram_ctrl.STATUS.SCR_KEY_SEED_VALID` bit should be set to 1.

@cfrantz @tjaychen @cdgori 
WDYT? Or is that too much complexity to add at this point in ROM code?

Note: The scrambling keys are in all cases ephemeral - the only difference is that the keyed KDF is using an all-zero value if the key seeds have not been provisioned yet.

**Comment by tjaychen** [comment]
i think it's probably not worth the complexity. As @msfschaffner said the keys are always ephemeral, and entropy information is always pulled to further enhance its creation.   If ROM arms the memory scrambling, we should just go with whatever we have.

The checking of whether the seed was valid IMO is fine to leave to later pieces of software if they care about it.

**Comment by msfschaffner** [comment]
Ok, in that case let's just ignore the value of `sram_ctrl.STATUS.SCR_KEY_SEED_VALID` in ROM code.

**Comment by mundaym** [comment]
Thanks, that's useful info, sorry for sidetracking the issue. The original proposal (access exceptions for SRAM accesses during scrambling) SGTM.

**Comment by msfschaffner** [comment]
We've discussed this in the security design sync today and decided to not change the currently implemented behavior. 

I.e., no TL-UL errors will be generated when the SRAM is accessed while a key request is pending. 
Rather, the SRAM just stalls the TL-UL request until the key has been obtained. 

Software may still choose to use the polling mechanism that we have discussed above, since that has already been implemented.
